### PR TITLE
Prevent PublisherSession.publish to throw ChannelClosed exception

### PR DESCRIPTION
**Pull request by sebwoj**
Fix for throwing ChannelClosed exception by PublisherSession.publish

### CODE REVIEW

**Comment by abompard** [comment]
Hey, thanks for reporting and fixing this issue. You've added another commit in this PR, which adds a new CLI subcommand. Could you keep that for another PR please, so we don't mix bugfixes and new features?
Also, do you mind writing a unit test for the raised exception please?
Thanks!

**Comment by abompard** [comment]
Thanks! Would you mind adding a unit test? It would probably be similar to `test_publish_reconnect_failed`.

**Comment by sebwoj** [comment]
OK. So I will write addition test for channelclosed exception. It will be the same as test_publish_reconnect_failed except second call to channel will throw channelclosed exception.
Do you think that renaming test_publish_reconnect_failed to test_publish_reconnect_failed_amqpconnectionerror and adding test_publish_reconnect_failed_channelclosed will be good idea? UT for my fix should be in the same commit with fix or in another?

**Comment by abompard** [comment]
You can rename the test, add yours, and do all that in a single commit, yes. Thanks

File: fedora_messaging/tests/unit/test_session.py [hunk 1]
```diff
@@ -195,6 +195,27 @@ def test_publish_reconnect_failed(self):
...
+        # close the connection if it can't be established.
+        self.publisher._channel.publish.side_effect = pika_errs.ConnectionClosed(
+            200, "I wanted to"
+        )
+        connection_class_mock = mock.Mock()
+        connection_mock = mock.Mock()
+        connection_class_mock.return_value = connection_mock
+        connection_mock.channel.side_effect = pika_errs.ChannelClosed(
+            404, "NOT_FOUND - no exchange 'amq.fake' in vhost '/'"
+        )
```
Comment (jeremycline) [inline fedora_messaging/tests/unit/test_session.py]: Mock objects return Mock objects by default, so I *believe* you can rewrite this as:

```
connection_class_mock = mock.Mock()
connection_class_mock.return_value.channel.side_effect = pika_errs.ChannelClosed(...)
```

It makes no functional change and it's a copy of existing code, but just an FYI :cake:

File: fedora_messaging/_session.py [hunk 2]
```diff
@@ -160,7 +160,7 @@ def publish(self, message, exchange=None):
             self._connection = self._channel = None
             try:
                 self._connect_and_publish(exchange, message)
-            except pika_errs.AMQPConnectionError as e:
+            except pika_errs.AMQPError as e:
```
[the remark this answers is not in the document]
Comment (sebwoj) [reply fedora_messaging/_session.py]: I agree. So inner try-catch should look like the outer one?
Do u think that information:
_log.warning("Message was rejected by the broker (%s)", str(e))
should be logged in the inner try-catch as it is in the outer, but with extra info like:
_log.warning("Reconnection: Message was rejected by the broker (%s)", str(e))
?

**Comment by sebwoj** [comment]
Why Codecov Report includes "removed" changes from cli?
Am I missed something?
Looks like that after my PR code coverage will be decreased?

File: fedora_messaging/_session.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremycline) [reply fedora_messaging/_session.py]: There should be a log message about resetting the connection and the message getting rejected probably won't have anything to do with that so I think the same log message is fine.

**Comment by jeremycline** [comment]
Something is wonky and it's not your fault. Travis uploads the coverage report to Codecov, so force pushing to get the CI to re-run *may* fix it, otherwise it's going to require some digging.

**Pull request merged**